File: problems/Edit Distance/minDistance.java
class Solution {
    public int minDistance(String word1, String word2) {
        Integer[][] dp = new Integer[word1.length()][word2.length()];

        return minDistanceTopDown(word1, word2, dp, 0, 0);
    }

    private int minDistanceTopDown(String word1, String word2, Integer[][] dp, int idx1, int idx2) {
        if (word1.length() == idx1) {
            return word2.length() - idx2;
        }

        if (word2.length() == idx2) {
            return word1.length() - idx1;
        }

        if (null == dp[idx1][idx2]) {
            if (word1.charAt(idx1) == word2.charAt(idx2)) {
                dp[idx1][idx2] = minDistanceTopDown(word1, word2, dp, 1 + idx1, 1 + idx2);
            } else {
                int del = 1 + minDistanceTopDown(word1, word2, dp, 1 + idx1, idx2);
                int ins = 1 + minDistanceTopDown(word1, word2, dp, idx1, 1 + idx2);
                int rep = 1 + minDistanceTopDown(word1, word2, dp, 1 + idx1, 1 + idx2);

                dp[idx1][idx2] = Math.min(del, Math.min(ins, rep));
            }
        }

        return dp[idx1][idx2];
    }

    private int minDistanceRecursive(String word1, String word2, int idx1, int idx2) {
        if (word1.length() == idx1) {
            return word2.length() - idx2;
        }

        if (word2.length() == idx2) {
            return word1.length() - idx1;
        }

        if (word1.charAt(idx1) == word2.charAt(idx2)) {
            return minDistanceRecursive(word1, word2, 1 + idx1, 1 + idx2);
        }

        int del = 1 + minDistanceRecursive(word1, word2, 1 + idx1, idx2);
        int ins = 1 + minDistanceRecursive(word1, word2, idx1, 1 + idx2);
        int rep = 1 + minDistanceRecursive(word1, word2, 1 + idx1, 1 + idx2);

        return Math.min(del, Math.min(ins, rep));
    }
}

File: problems/Edit Distance/minDistance.cs
public class Solution {
    public int MinDistance(string word1, string word2) {
        return minDistanceBottomUp(word1, word2);
    }

    private int minDistanceBottomUp(string word1, string word2) {
        int len1 = word1.Length;
        int len2 = word2.Length;
        int[,] dp = new int[1 + len1, 1 + len2];

        for (int idx1 = 0; len1 >= idx1; ++idx1) {
            for (int idx2 = 0; len2 >= idx2; ++idx2) {
                if (0 == idx1) {
                    dp[0, idx2] = idx2;
                } else if (0 == idx2) {
                    dp[idx1, 0] = idx1;
                } else if (word1[idx1 - 1] == word2[idx2 - 1]) {
                    dp[idx1, idx2] = dp[idx1 - 1, idx2 - 1];
                } else {
                    int delete = dp[idx1 - 1, idx2];
                    int insert = dp[idx1, idx2 - 1];
                    int replace = dp[idx1 - 1, idx2 - 1];

                    dp[idx1, idx2] = 1 + Math.Min(delete, Math.Min(insert, replace));
                }
            }
        }

        return dp[len1, len2];
    }
}
